run.sh

```diff
@@ -1,69 +1,94 @@
+print_help() {
+  printf "Help: \n \
+        -v) Verbose output.\n \
+        -m) Test mode. [ all | loopback | ossl_server | ossl_client | bssl_server | bssl_shim ]\n \
+        -c) Case file. (Optional. Default cases.json) \n"
 }
 
-run_boring_client() {
-  cargo run -- \
-  --client $BASE_DIR/../boringssl/build/ssl/test/bssl_shim \
-  --server $BASE_DIR/../dist/Debug/bin/nss_bogo_shim \
-  --rootdir $BASE_DIR/../boringssl/ssl/test/runner/ \
-  --test-cases $BASE_DIR/$CASE_FILE
-}
+while [ $# -gt 0 ]; do
+    case $1 in
+        -v) export RUST_LOG=debug ;;
+        -m) MODE="$2"; shift ;;
+        -c) CASE_FILE="$2"; shift;;
+        *) print_help; exit 2 ;;
+    esac
+    shift
+done
 
-run_ossl_server() {
-  cargo run -- \
-  --client $BASE_DIR/../dist/Debug/bin/nss_bogo_shim \
-  --server $BASE_DIR/../openssl/test/ossl_shim/ossl_shim \
-  --rootdir $BASE_DIR/../boringssl/ssl/test/runner/ \
-  --test-cases $BASE_DIR/$CASE_FILE \
-  --client-writes-first \
-  --force-IPv4
-}
+CERT_DIR="$BASE_DIR/../boringssl/ssl/test/runner/"
+NSS_SHIM="$BASE_DIR/../dist/Debug/bin/nss_bogo_shim"
+BSSL_SHIM="$BASE_DIR/../boringssl/build/ssl/test/bssl_shim"
+OSSL_SHIM="$BASE_DIR/../openssl/test/ossl_shim/ossl_shim"
+
+SHIM_ARRAY=($NSS_SHIM $BSSL_SHIM $OSSL_SHIM)
+
+run_shim_pair() {
+  SHIM1=$1
+  SHIM2=$2
+  # If NSS acts as the client, interop needs this argument.
+  # It would become obsolete if bssl and ossl could actively initiate
+  # communication after the handshake.
+  CLIENT_WRITES=""
+  if [[ $SHIM1 = *"nss_bogo_shim"* ]]; then
+    CLIENT_WRITES="--client-writes-first"
+  fi
+
+  # The ossl_shim is currently not properly IPv6 capable, which is why interop
+  # needs this argument when ossl_shim is involved in the test case.
+  IP4=""
+  if [[ $SHIM1 = *"ossl"* ]] || [[ $SHIM2 = *"ossl"* ]] ; then
+    IP4="--force-IPv4"
+  fi
 
-run_ossl_client() {
   cargo run -- \
-  --client $BASE_DIR/../openssl/test/ossl_shim/ossl_shim \
-  --server $BASE_DIR/../dist/Debug/bin/nss_bogo_shim \
-  --rootdir $BASE_DIR/../boringssl/ssl/test/runner/ \
+  --client $SHIM1 \
+  --server $SHIM2 \
+  --rootdir $CERT_DIR \
   --test-cases $BASE_DIR/$CASE_FILE \
-  --force-IPv4
+  $CLIENT_WRITES \
+  $IP4
 }
 
-run_loopback() {
-  cargo run -- \
-  --client $BASE_DIR/../dist/Debug/bin/nss_bogo_shim \
-  --server $BASE_DIR/../dist/Debug/bin/nss_bogo_shim \
-  --rootdir $BASE_DIR/../boringssl/ssl/test/runner/ \
-  --test-cases $BASE_DIR/$CASE_FILE
-}
+case $MODE in
+  "all")
+      for i in ${SHIM_ARRAY[@]}
+        do
+          for j in ${SHIM_ARRAY[@]}
+          do
+            # Currently at least one nss shim needs to be involved in a test
+            # case because neither bssl_shim nor ossl_shim can actively
+            # initiate the communication after a successful handshake.
+            if [[ $i = *"nss_bogo_shim"* ]] || [[ $j = *"nss_bogo_shim"* ]] ; then
+              run_shim_pair $i $j
+            fi
+          done
+        done
+      ;;
```

Comment:
At this point, it might be easier to just run through all the options.

e.g.,

```
run_mode() {
  if [[ "$1" == "$2" ]] || [[ "$1" == "all" ]]; then
    run_shim_pair "$3" "$4"
    invalid_mode=
  fi
}
invalid_mode=true
run_mode "$MODE" boring_server "$NSS_SHIM" "$BSSL_SHIM"
run_mode "$MODE" boring_client "$BSSL_SHIM" "$NSS_SHIM"
# etc...
[[ -z "$invalid_mode" ]] || print_help
```

Reply:
That is indeed more elegant than my solution. I adapted yours.